File: leetsync/2562-count-ways-to-build-good-strings/count-ways-to-build-good-strings.go
func countGoodStrings(low int, high int, zero int, one int) int {
	const mod = 1_000_000_007
	dp := make([]int, high+1)
	dp[0] = 1
	for i := 1; i < len(dp); i++ {
		if i >= zero {
			dp[i] += dp[i-zero]
		}
		if i >= one {
			dp[i] += dp[i-one]
		}
		dp[i] %= mod
	}
	result := 0
	for i := low; i <= high; i++ {
		result += dp[i]
		result %= mod
	}
	return result
}

File: leetsync/2562-count-ways-to-build-good-strings/count-ways-to-build-good-strings.java
class Solution {
    int mod = 1000000007;
    int[] dp;

    public int countGoodStrings(int low, int high, int zero, int one) {
        dp = new int[1000001];
        Arrays.fill(dp, -1);
        return solveRecursionMemoization(low, high, zero, one, 0);
    }

    private int solveRecursionMemoization(int low, int high, int zero, int one, int currentStringSize) {
        if (currentStringSize > high) {
            return 0;
        }
        if (dp[currentStringSize] != -1) {
            return dp[currentStringSize];
        }
        int addOne = 0;
        if (currentStringSize >= low && currentStringSize <= high) {
            addOne = 1;
        }
        int appendZero = solveRecursionMemoization(low, high, zero, one, currentStringSize + zero);
        int appendOne = solveRecursionMemoization(low, high, zero, one, currentStringSize + one);
        return dp[currentStringSize] = (addOne + appendZero + appendOne) % mod;
    }
}